python/cuml/cuml/metrics/regression.pxd

```diff
@@ -1,29 +0,0 @@
-#
-# Copyright (c) 2019-2022, NVIDIA CORPORATION.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#
-
-from pylibraft.common.handle cimport handle_t
-
-cdef extern from "cuml/metrics/metrics.hpp" namespace "ML::Metrics":
-
-    float r2_score_py(const handle_t& handle,
```

Comment:
Why are we switching to another implementation when we have a C++ one? Unless there is functionality or correctness missing here, I don't see the need, and in general moves us away from matching C++ and python functionality

Reply:
This was covered in my top-level comment. The `cupy` path for the other regression score functions has been there, untouched, for ~4 years. This just moves the `r2_score` function to do the same. It brings us into compatibility with the scikit-learn implementation (adding support for several previously unsupported code keyword arguments), and lets us fix some 0cc issues. If later on we want to move this functionality back to C++ I think that's fine, but I don't see why a move to use cupy in a place we already use cupy for other metrics is a problem.

Reply:
I don't believe `r2_score` (or any of the scoring functions) are on the performance-critical path for most use cases. Taking the simplest path to ensuring we can match sklearn functionality here seems the correct engineering choice - as Corey said above, "we implement in cupy for convenience and will move things to c++ later / as needed". I don't believe C++ is needed here at the moment, and using cupy is certainly more convenient.

Reply:
I think that's a fair justification, particularly multioutput is easier. As far as I'm aware, we don't have active users for the C++ API of r2_score currently. It should also be relatively quickly to port to C++ when needed, though it doesn't seem like the highest priority as you point out since we haven't done it in 4 years. 

This does beg the question if at some point the logical migration will be to use scikit-learn's r2_score with array_api GPU inputs, which I think it already supports ( https://github.com/scikit-learn/scikit-learn/blob/6a0838c416c7c2a6ee7fe4562cd34ae133674b2e/sklearn/metrics/_regression.py#L1275) but @betatim would know better

Reply:
It does, but not by default (yet). Not sure when they'll enable the array api by default, but I agree if/when that happens we could just use their implementation.

Reply:
I think the counter argument would be that the C++ code itself is not particularly complex and located in RAFT: https://github.com/rapidsai/raft/blob/68d412a119f918dc003ecd0c430b6bde7499bc96/cpp/include/raft/stats/detail/scores.cuh#L58. This means that the code can be used by other C++ codebases and we have a single source of code to mantain between C++ and Python. 

Introducing a new python code means we end up with both implementations, but raises the question of the fact that right now the only C++/python codebase that uses r2_score particularly is cuML, I don't think cuVS or cuGraph would use it. @cjnolet would love your input here if you see something like r2_score worth the maintenance in RAFT _if_ it'll only be used in cuML.

Reply:
I'd argue that YAGNI. If we're the only current consumer of `r2_score` in RAFT, and we know of no users using the `py_r2_score` exposed in `libcuml`, then the location and kind of code is purely an implementation detail that no user (currently) cares about. For `cuml` the thing we care about here is matching sklearn's interface, and being able to easily amend that interface as the sklearn api evolves. Having the python code call c++ code in `libcuml` which calls a shared implementation in `raft` adds unjustified (at this time) complexity IMO.

Reply:
We also need to resolve it in collaboration with @cjnolet who is the person in charge of RAFT, and a reviewer of the PR.

Reply:
@dantegd I was strictly referring to the libcuml API here, not RAFT.

Reply:
> so that question is essentially the same

I will stand by that the decision whether we maintain this function as part of the libcuml API should be a question of utilization and performance and is in principle immaterial to the RAFT API.